records/2022-04-14.secrets-usage.md

```diff
@@ -0,0 +1,154 @@
+# Secrets in Notify
+
+Date: 2022-04-14
+
+## Status
+
+**DRAFT**.
+
+## Context
+
+With potential secrets leaked to the public, we need exhaustive documentation that
+provides a quick reference on how secrets in Notify are used. This document aims to
+explain how these interact with different components of our application.
+
+## Methods and libraries
+
+This section enumerates a few methods and libraries to help understand the rest of
+the document to casual readers.
+
+### [itsdangerous](https://itsdangerous.palletsprojects.com/)
+
+This is a library that specializes in signing content in a cryptographic manner using
+a [HMAC-SHA1 algorithm](https://en.wikipedia.org/wiki/HMAC). This will not encrypt a
+message in a one-way manner; we can reverse the operation even without the secret and
+salt used throughout the signing process. Rather, the goal is to guarantee
+authenticity and integrity of a message.
+
+> In cryptography, an HMAC (sometimes expanded as either keyed-hash message authentication code or hash-based message authentication code) is a specific type of message authentication code (MAC) involving a cryptographic hash function and a secret cryptographic key. As with any MAC, it may be used to simultaneously verify both the data integrity and authenticity of a message.
+
+> HMAC can provide authentication using a shared secret instead of using digital signatures with asymmetric cryptography. It trades off the need for a complex public key infrastructure by delegating the key exchange to the communicating parties, who are responsible for establishing and using a trusted channel to agree on the key prior to communication.
+
+> [Wikipedia](https://en.wikipedia.org/wiki/HMAC)
+
+#### "Decrypting" content signed by itsdangerous
+
+Content signed with the [itsdangerous](https://itsdangerous.palletsprojects.com/)
+library is merely compressed with zlib (sometimes if above an arbitrary size limit)
+and base64 encoded. To "decrypt" signed content, perform the following to revert
+to previous readable format without using the secret and salt:
+
+```python
+import zlib
+import base64
+
+p = '<paste personalized data here>'
+b64 = p[1:]
+b64_utf = b64.encode('utf-8')
+b64_decoded = base64.urlsafe_b64decode(b64_utf + b'==')
+print(zlib.decompress(b64_decoded))
```

Comment:
We can also get the unsigned content by using `URLSafeTimedSerializer("").loads_unsafe(p)[1]`

Reply:
Ah maybe, it's possible that itsdangerous offers the utility to do it. I just tried it locally though and that returned `(False, None)` on a value where the algo above worked. We can debug tomorrow for curiosity sake. 

I find it nice to be able to decode the value without any help from itsdangerous, to show what's done with no magic. We can include that line for an easier way to do it too.